lib/graphql/analysis/ast/query_complexity.rb

```diff
@@ -8,113 +8,211 @@ class QueryComplexity < Analyzer
+          # but I think that's ok. If the arguments _didn't_ match,
+          # then the query would have been rejected as invalid.
+          complexities_on_type = @complexities_on_type_by_query[visitor.query] ||= [ScopedTypeComplexity.new(nil, nil, query)]
+
+          complexity = complexities_on_type.last.scoped_children[parent_type][field_key] ||= ScopedTypeComplexity.new(node, visitor.field_definition, visitor.query)
+          # Push it on the stack.
+          complexities_on_type.push(complexity)
         end
 
         def on_leave_field(node, parent, visitor)
           # We don't want to visit fragment definitions,
           # we'll visit them when we hit the spreads instead
           return if visitor.visiting_fragment_definition?
           return if visitor.skipping?
-
-          type_complexities = @complexities_on_type.pop
-          child_complexity = type_complexities.max_possible_complexity
-          own_complexity = get_complexity(node, visitor.field_definition, child_complexity, visitor)
-
-          if @complexities_on_type.last.is_a?(AbstractTypeComplexity)
-            key = selection_key(visitor.response_path, visitor.query)
-            parent_type = visitor.parent_type_definition
-            visitor.query.possible_types(parent_type).each do |type|
-              @complexities_on_type.last.merge(type, key, own_complexity)
-            end
-          else
-            @complexities_on_type.last.merge(own_complexity)
-          end
+          complexities_on_type = @complexities_on_type_by_query[visitor.query] ||= [ScopedTypeComplexity.new(nil, nil, query)]
+          complexities_on_type.pop
         end
 
         # @return [Integer]
         def max_possible_complexity
-          @complexities_on_type.last.max_possible_complexity
-        end
-
-        private
-
-        def selection_key(response_path, query)
-          # We add the query object id to support multiplex queries
-          # even if they have the same response path, they should
-          # always be added.
-          "#{response_path.join(".")}-#{query.object_id}"
-        end
-
-        # Get a complexity value for a field,
-        # by getting the number or calling its proc
-        def get_complexity(ast_node, field_defn, child_complexity, visitor)
-        # Return if we've visited this response path before (not counting duplicates)
-          defined_complexity = field_defn.complexity
-
-          arguments = visitor.arguments_for(ast_node, field_defn)
-
-          case defined_complexity
-          when Proc
-            defined_complexity.call(visitor.query.context, arguments, child_complexity)
-          when Numeric
-            defined_complexity + (child_complexity || 0)
-          else
-            raise("Invalid complexity: #{defined_complexity.inspect} on #{field_defn.name}")
+          val = 0
+          @complexities_on_type_by_query.each do |query, complexities_on_type|
```

Comment:
Maybe use `.reduce(0) do |sum, (query, complexities_on_type)| do` here instead of .each with an initial value declared above?

Reply:
:+1: